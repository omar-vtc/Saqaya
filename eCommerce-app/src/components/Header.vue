<template>
  <v-app-bar :elevation="2">
    <!-- Left-side menu (dropdown on small screens) -->
    <div class="d-flex d-md-none">
      <v-menu>
        <template #activator="{ props }">
          <v-app-bar-nav-icon v-bind="props" />
        </template>

        <v-list>
          <v-list-item :to="'/products'" title="Products" />
          <v-list-item :to="'/contact-us'" title="Contact Us" />
          <v-list-item :to="'/cart'" title="Cart" />
        </v-list>
      </v-menu>
    </div>

    <!-- Keep nav icon as static icon on md+ (optional) -->
    <div class="d-none d-md-flex">
      <v-app-bar-nav-icon />
    </div>

    <!-- App title -->
    <div style="max-width: 120px">
      <v-app-bar-title class="text-left">SaQaya</v-app-bar-title>
    </div>
    <!-- Center nav links on md+ -->
    <v-spacer />
    <div class="d-none d-md-flex align-center">
      <v-btn text :to="'/home'" class="mx-2">Home</v-btn>

      <v-btn text :to="'/products'" class="mx-2">Products</v-btn>
      <v-btn text :to="'/contact-us'" class="mx-2">Contact Us</v-btn>
    </div>
    <v-spacer />

    <!-- Right-side icons -->
    <template v-slot:append>
      <v-btn icon="mdi-cart" :to="'/cart'" />
      <v-btn icon="mdi-magnify" />
    </template>
  </v-app-bar>
</template>
