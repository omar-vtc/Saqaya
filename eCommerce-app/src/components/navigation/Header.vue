<template>
  <div class="header">
    <div class="header__logo">
      <img src="../../assets//img/logo2.png" alt="" />
    </div>
    <div class="header__list-container">
      <ul class="header__list">
        <li class="header__list-item"><p class="list-item__ele">Home</p></li>
        <li class="header__list-item">
          <p class="list-item__ele">Products</p>
        </li>
        <li class="header__list-item">
          <p class="list-item__ele">Contact Us</p>
        </li>
      </ul>
    </div>
    <div class="header__actions">actions div</div>
  </div>
</template>

<style scoped>
* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

.header {
  width: 100%;
  height: 6rem;
  border: solid;
  display: flex;
  justify-content: space-between;
  padding: 0 2rem;
  background-color: #ffffff;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.05);
}

.header__logo {
  flex: 1;
  height: inherit;
  border: solid red;
  display: flex;
  justify-content: center;
}

.header__list-container {
  flex: 5;
  height: inherit;
  border: solid blue;
  display: flex;
  align-items: center;
  /* justify-content: center; */
}

.header__actions {
  flex: 1;
  height: inherit;
  border: solid green;
}

.header__list {
  list-style: none;
  display: flex;
  gap: 2.5rem;
  border: solid;
}

.header__list-item .list-item__ele {
  font-size: 1.2rem;
  font-weight: 500;
  color: #333;
  cursor: pointer;
  transition: color 0.2s ease;
  min-width: max-content;
  padding: 0.5rem 1.2rem;
  width: max-content;
  /* border: solid green; */
  border-radius: 7px;
}

.header__list-item .list-item__ele:hover {
  color: #ff00ae;
  background-color: #6e6d6d1c;
}

.header__logo img {
  border: solid;
  object-fit: contain;
}
</style>
