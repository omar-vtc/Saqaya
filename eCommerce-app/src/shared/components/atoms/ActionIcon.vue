<script setup lang="ts">
import { defineProps } from "vue";

const props = defineProps<{
  className?: string;
  IconClass: string | [string, string];
  onClick?: () => void; // optional click handler
}>();
// console.log(props.IconClass);
</script>

<template>
  <div :class="props.className">
    <font-awesome-icon :icon="props.IconClass" @click="props.onClick" />
  </div>
</template>
