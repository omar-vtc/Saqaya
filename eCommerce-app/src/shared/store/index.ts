// store/index.ts
import { createStore } from "vuex";
import { products, type ProductsState } from "./modules/products";
import { cart, type CartState } from "./modules/cart";
import createPersistedState from "vuex-persistedstate";

export interface RootState {
  products: ProductsState;
  cart: CartState;
}

export const store = createStore<RootState>({
  modules: {
    products,
    cart,
  },
  plugins: [
    createPersistedState({
      paths: ["cart"], // only persist the cart module
    }),
  ],
});
