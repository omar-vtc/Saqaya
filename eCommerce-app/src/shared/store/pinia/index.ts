// stores/index.ts

export * from "./modules/cart";
export * from "./modules/products";
