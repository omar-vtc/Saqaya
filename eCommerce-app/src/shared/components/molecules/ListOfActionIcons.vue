<script setup lang="ts">
import { defineProps } from "vue";
import ActionIcon from "../atoms/ActionIcon.vue";

type IconItem = {
  name: string | [string, string]; // FontAwesome icon name
  onClick?: () => void; // optional click handler
};

const props = defineProps<{
  className?: string;
  iconClass?: string;
  icons: IconItem[];
}>();
</script>

<template>
  <div :class="props.className">
    <ActionIcon
      v-for="(icon, index) in props.icons"
      :key="index"
      :class-name="props.iconClass"
      :IconClass="icon.name"
      @click="icon.onClick"
    />
  </div>
</template>
