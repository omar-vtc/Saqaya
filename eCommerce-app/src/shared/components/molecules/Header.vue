<script lang="ts">
import { ref, computed } from "vue";
import { mapGetters, mapActions } from "vuex";
import Logo from "../atoms/Logo.vue";
import NavList from "./NavList.vue";
import ListOfActionIcons from "./ListOfActionIcons.vue";
import ActionIcon from "../atoms/ActionIcon.vue";
import Cart from "./Cart.vue";

export default {
  components: {
    Logo,
    NavList,
    ListOfActionIcons,
    ActionIcon,
    Cart,
  },
  data() {
    return {
      isDrawerOpen: false,
      isCartOpen: false,
    };
  },
  computed: {
    ...mapGetters("cart", ["cartItems"]),
  },
  methods: {
    ...mapActions("cart", ["removeProductFromCart"]),
    toggleDrawer() {
      if (this.isCartOpen) this.isCartOpen = false;
      this.isDrawerOpen = !this.isDrawerOpen;
    },
    toggleCartDrawer() {
      if (this.isDrawerOpen) this.isDrawerOpen = false;
      this.isCartOpen = !this.isCartOpen;
    },
    handleRemove(id: number) {
      this.removeProductFromCart(id);
    },
  },
};
</script>

<template>
  <div class="header">
    <Logo class-name="header__logo" img-class-name="header__logo--img" />
    <!-- Hamburger icon for mobile -->
    <ActionIcon
      className="header__hamburger"
      IconClass="bars"
      @Click="toggleDrawer"
    />

    <!-- Main navigation (hidden on small screens) -->
    <NavList
      div-class-name="header__list-container"
      class-name="header__list"
      ele-class-name="header__list-item"
    />

    <!-- Action Icons -->
    <ListOfActionIcons
      className="header__actions"
      iconClass="action-icon"
      :icons="[
        { name: 'magnifying-glass', 'data-testid': 'icon-magnifying-glass' },
        {
          name: 'shopping-cart',
          onClick: toggleCartDrawer,
          badgeCount: cartItems.length,
          'data-testid': 'icon-shopping-cart',
        },
        { name: 'right-to-bracket', 'data-testid': 'icon-login' },
      ]"
    />
  </div>

  <!-- Drawer on small screens -->
  <transition name="mobile-slide">
    <div class="mobile-drawer" v-if="isDrawerOpen">
      <!-- Logo -->
      <Logo
        class-name="mobile-drawer__logo"
        img-class-name="mobile-drawer__logo--img"
      />

      <!-- drawer Nav Links -->

      <NavList div-class-name="mobile-drawer__list" class-name="header__list"
      ele-class-name="header__list-item" onClick: toggleCartDrawer />

      <!-- Actions -->
      <ListOfActionIcons
        className="mobile-drawer__actions"
        iconClass="action-icon"
        :icons="[
          { name: 'magnifying-glass' },
          { name: 'shopping-cart', onClick: toggleCartDrawer },
          { name: 'right-to-bracket' },
        ]"
      />
    </div>
  </transition>

  <!-- Cart Drawer -->
  <Cart :isOpen="isCartOpen" @close="toggleCartDrawer" />
</template>

<style>
* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

.list-item__ele.router-link-exact-active,
.list-item__ele.router-link-active {
  background-color: #b6b5b510;
  color: #ff00ae !important;
}

.header {
  width: 100%;
  height: 6rem;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0 2rem;
  background-color: #ffffff;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.05);
  position: relative;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  z-index: 1000;
}

.header__logo {
  flex: 1;
  height: inherit;
  display: flex;
  align-items: center;
  justify-content: center;
  /* border: solid; */
}

.header__logo--img {
  object-fit: contain;
  height: 95%;
  max-width: 150px;
}

.header__hamburger {
  display: none;
  font-size: 1.5rem;
  cursor: pointer;
  color: #333;
}

.header__list-container {
  flex: 5;
  height: inherit;
  display: flex;
  align-items: center;
}

.header__list {
  list-style: none;
  display: flex;
  gap: 2.5rem;
}

.header__list-item .list-item__ele {
  font-size: 1.2rem;
  font-weight: 500;
  color: #333;
  cursor: pointer;
  transition: color 0.2s ease;
  min-width: max-content;
  padding: 0.5rem 1.2rem;
  width: max-content;
  border-radius: 7px;
  text-decoration: none;
}

.header__list-item .list-item__ele:hover {
  color: #ff00ae;
  background-color: #b6b5b510;
}

.header__actions {
  flex: 1;
  height: inherit;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 1.2rem;
}

.action-icon {
  font-size: 1.5rem;
  font-weight: 500;
  color: #333;
  cursor: pointer;
  transition: background-color 0.2s ease, color 0.2s ease;
}

.action-icon:hover {
  color: #ff00ae;
}
.icon-container {
  position: relative;
  display: inline-block;
}
.icon-badge {
  position: absolute;
  top: -0.4rem;
  right: -0.6rem;
  background-color: #ff008c;
  color: white;
  font-size: 0.65rem;
  font-weight: bold;
  padding: 0.15rem 0.4rem;
  border-radius: 50%;
  line-height: 1;
}
/* Slide transition */
.slide-enter-active,
.slide-leave-active {
  transition: transform 0.3s ease, opacity 0.3s ease;
}
.slide-enter-from,
.slide-leave-to {
  transform: translateX(100%);
  opacity: 0;
}

/* Mobile Drawer */

/* Slide in from left (enter) */
.mobile-slide-enter-active {
  transition: transform 0.3s ease, opacity 0.3s ease;
}
.mobile-slide-enter-from {
  transform: translateX(-100%);
  opacity: 0;
}
.mobile-slide-enter-to {
  transform: translateX(0%);
  opacity: 1;
}

/* Slide out to right (leave) */
.mobile-slide-leave-active {
  transition: transform 0.3s ease, opacity 0.3s ease;
}
.mobile-slide-leave-from {
  transform: translateX(0%);
  opacity: 1;
}
.mobile-slide-leave-to {
  transform: translateX(-100%);
  opacity: 0;
}
.mobile-drawer {
  position: fixed;
  top: 0;
  left: 0;
  height: 100%;
  width: 75%;
  max-width: 300px;
  background-color: white;
  box-shadow: 2px 0 8px rgba(0, 0, 0, 0.1);
  padding-top: 6rem;
  z-index: 999;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}

.mobile-drawer__logo {
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 1rem 0;
  border-bottom: 1px solid #eee;
}

.mobile-drawer__logo--img {
  height: 40px;
  object-fit: contain;
}

.mobile-drawer__list {
  list-style: none;
  padding: 1rem 0;
  margin: 0;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: 1rem;
  padding-left: 2rem;
  flex: 1;
}

.mobile-drawer__list--listItem {
  font-size: 1.2rem;
  font-weight: 500;
  color: #333;
  padding: 0.5rem 1.5rem;
  transition: color 0.2s ease;
}

.mobile-drawer__list--listItem:hover {
  color: #ff00ae;
}

.mobile-drawer__actions {
  display: flex;
  justify-content: center;
  gap: 1.2rem;
  padding: 1rem 0;
  border-top: 1px solid #eee;
}

/* Cart Drawer */
.cart-drawer {
  position: fixed;
  top: 6rem;
  right: 0;
  height: 90%;
  width: 300px;
  background-color: white;
  box-shadow: -2px 0 8px rgba(0, 0, 0, 0.1);
  z-index: 999;
  display: flex;
  flex-direction: column;
  padding: 1rem;
  /* border: solid black; */
}

.cart-drawer__header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-bottom: 1px solid #eee;
  padding-bottom: 1rem;
}

.cart-drawer__header h3 {
  font-size: 1.3rem;
  margin: 0;
}

.close-btn {
  font-size: 1.5rem;
  cursor: pointer;
}

.cart-drawer__content {
  flex: 1;
  padding-top: 1rem;
  overflow-y: auto;
}

/*cart content*/
.cart-product-card {
  display: flex;
  gap: 0.8rem;
  margin-bottom: 1rem;
  padding-bottom: 1rem;
  border-bottom: 1px solid #eee;
}

.cart-product-card__image {
  width: 70px;
  height: 70px;
  object-fit: cover;
  border-radius: 6px;
}

.cart-product-card__info {
  display: flex;
  flex-direction: column;
  justify-content: center;
  flex: 1;
}

.cart-product-card__title {
  font-size: 1rem;
  font-weight: 500;
  margin: 0;
}

.cart-product-card__price {
  font-size: 0.95rem;
  color: #610243;
  margin: 0.25rem 0;
}

.cart-product-card__remove {
  background: none;
  border: none;
  color: #ff0040;
  font-size: 0.85rem;
  cursor: pointer;
  padding: 0;
  text-align: left;
}
.cart-drawer__footer {
  border-top: 1px solid #eee;
  padding-top: 1rem;
  margin-top: 1rem;
  display: flex;
  justify-content: space-between;
  font-weight: 600;
  font-size: 1rem;
}
.cart-total {
  display: flex;
  justify-content: space-between;
  font-weight: 600;
  font-size: 1rem;
  align-items: center;
}

.cart-total__amount {
  color: #610243;
}
.checkout-button {
  padding: 0.5rem 0.5rem;
  background-color: #610243;
  color: #fff;
  border: none;
  border-radius: 5px;
  font-size: 1rem;
  font-weight: 500;
  cursor: pointer;
  transition: background-color 0.3s ease;
  margin-left: 20px;
}

.checkout-button:hover {
  background-color: #8a055d;
}
/* Responsive */
@media (max-width: 922px) {
  .header {
    padding: 0 1rem;
  }

  .header__list {
    gap: 1.5rem;
    /* flex-direction: column; */
  }

  .header__list-item .list-item__ele {
    font-size: 1rem;
    padding: 0.4rem 0.8rem;
  }

  .header__logo img {
    max-width: 70px;
  }

  .action-icon {
    font-size: 1.3rem;
  }
}

/* Mobile vertical layout */
@media (max-width: 600px) {
  .header__list {
    gap: 1.5rem;
    flex-direction: column;
  }
  .mobile-drawer__list {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    gap: 1rem;
    padding-left: 2rem;
  }

  .mobile-drawer__list .list-item__ele {
    padding: 0.5rem 1.5rem;
    width: 100%;
    text-align: left;
  }
  .header__logo img {
    max-width: 70px;
    display: none;
  }
  .header__actions {
    display: none;
  }
}

@media (max-width: 600px) {
  .header {
    flex-wrap: wrap;
    height: auto;
    gap: 1rem;
    padding: 1rem;
  }

  .header__hamburger {
    display: block;
  }

  .header__list-container {
    display: none;
  }

  .header__logo {
    flex: none;
    justify-content: center;
    width: 100%;
  }

  .header__actions {
    flex: none;
    justify-content: center;
    width: 100%;
  }

  .action-icon {
    font-size: 1.2rem;
  }
}

@media (min-width: 601px) {
  .header__hamburger {
    display: none !important;
  }
}
</style>
