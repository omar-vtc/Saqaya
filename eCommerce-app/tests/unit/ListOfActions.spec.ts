/// <reference types="vitest" />
import { describe, it, expect, beforeEach } from "vitest";
import { mount } from "@vue/test-utils";
import { createPinia, setActivePinia } from "pinia";
import { useCartStore } from "../../src/shared/store/pinia"; // adjust path to your Pinia store
import Header from "../../src/shared/components/molecules/Header.vue";
import ListOfActionIcons from "../../src/shared/components/molecules/ListOfActionIcons.vue";
import ActionIcon from "../../src/shared/components/atoms/ActionIcon.vue";

// Mock product type
type Product = {
  id: number;
  name: string;
  price: number;
  image: string;
};

// Helper to create a fresh pinia instance and prefill the cart
const mountHeaderWithCart = (items: Product[]) => {
  const pinia = createPinia();
  setActivePinia(pinia);
  const cartStore = useCartStore();
  cartStore.$patch({ items });

  return mount(Header, {
    global: {
      plugins: [pinia],
      components: {
        ActionIcon,
        ListOfActionIcons,
      },
      stubs: {
        "font-awesome-icon": true,
        "router-link": true,
      },
    },
  });
};

describe("Header.vue", () => {
  it("renders core components", () => {
    const wrapper = mountHeaderWithCart([]);

    expect(wrapper.findComponent({ name: "Logo" }).exists()).toBe(true);
    expect(wrapper.findComponent({ name: "NavList" }).exists()).toBe(true);

    const icons = wrapper.findAllComponents(ActionIcon);
    const iconProps = icons.map((el) => el.props("IconClass"));
    expect(iconProps).toEqual(
      expect.arrayContaining(["magnifying-glass", "shopping-cart"])
    );
  });

  it("toggles cart drawer when shopping-cart icon is clicked", async () => {
    const wrapper = mountHeaderWithCart([
      { id: 1, name: "Product 1", price: 100, image: "test.jpg" },
    ]);

    const cartButton = wrapper.find('[data-testid="icon-shopping-cart"]');
    expect(cartButton.exists()).toBe(true);

    await cartButton.trigger("click");
    await wrapper.vm.$nextTick();

    expect(wrapper.find(".cart-drawer").exists()).toBe(true);
  });

  it("displays correct cart item count in badge", () => {
    const wrapper = mountHeaderWithCart([
      { id: 1, name: "P1", price: 10, image: "" },
      { id: 2, name: "P2", price: 20, image: "" },
      { id: 3, name: "P3", price: 30, image: "" },
    ]);

    const badge = wrapper.find(
      '[data-testid="icon-shopping-cart"] .icon-badge'
    );
    expect(badge.exists()).toBe(true);
    expect(badge.text()).toBe("3");
  });
});
