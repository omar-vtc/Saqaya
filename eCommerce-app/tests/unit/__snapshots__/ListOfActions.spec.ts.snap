// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListOfActionIcons.vue > matches snapshot 1`] = `
"<div class="">
  <div class="icon-home icon-container">
    <font-awesome-icon-stub icon="fa fa-home"></font-awesome-icon-stub><span class="icon-badge">2</span>
  </div>
  <div class="icon-cart icon-container">
    <font-awesome-icon-stub icon="fas,cart-plus"></font-awesome-icon-stub><span class="icon-badge">5</span>
  </div>
</div>"
`;
